sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h

```diff
@@ -63,29 +54,60 @@ class AsyncMetricStorage : public MetricStorage
         cumulative_hash_map_->Set(measurement.first,
                                   DefaultAggregation::CloneAggregation(
                                       aggregation_type_, instrument_descriptor_, *delta));
-        delta_hash_map->Set(measurement.first, std::move(delta));
+        delta_hash_map_->Set(measurement.first, std::move(delta));
       }
       else
       {
         cumulative_hash_map_->Set(
             measurement.first,
             DefaultAggregation::CloneAggregation(aggregation_type_, instrument_descriptor_, *aggr));
-        delta_hash_map->Set(measurement.first, std::move(aggr));
+        delta_hash_map_->Set(measurement.first, std::move(aggr));
       }
     }
+  }
+
+  void RecordLong(
+      const std::unordered_map<MetricAttributes, long, AttributeHashGenerator> &measurements,
+      opentelemetry::common::SystemTimestamp observation_time) noexcept override
+  {
+    if (instrument_descriptor_.value_type_ != InstrumentValueType::kLong)
+    {
+      return;
+    }
+    Record<long>(measurements, observation_time);
+  }
+
+  void RecordDouble(
+      const std::unordered_map<MetricAttributes, double, AttributeHashGenerator> &measurements,
+      opentelemetry::common::SystemTimestamp observation_time) noexcept override
+  {
+    if (instrument_descriptor_.value_type_ != InstrumentValueType::kDouble)
+    {
+      return;
+    }
+    Record<double>(measurements, observation_time);
+  }
+
+  bool Collect(CollectorHandle *collector,
+               nostd::span<std::shared_ptr<CollectorHandle>> collectors,
+               opentelemetry::common::SystemTimestamp sdk_start_ts,
+               opentelemetry::common::SystemTimestamp collection_ts,
+               nostd::function_ref<bool(MetricData)> metric_collection_callback) noexcept override
+  {
 
     return temporal_metric_storage_.buildMetrics(collector, collectors, sdk_start_ts, collection_ts,
-                                                 std::move(delta_hash_map),
+                                                 std::move(delta_hash_map_),
```

Comment:
can we have `Record<T>` calls after `Collect`, if yes `delta_hash_map_` should not be accessed?

Reply:
Sorry the suggestion is not clear to me.
There would be multiple Record<T> calls for each of the registered callbacks for the storage, and then the Collect() would be invoked.

Reply:
I mean are we sure that `delta_hash_map_` is not accessed after  move?

Reply:
Oh nice catch. Yes, it would be called after `move`. The usual flow would be  
[Record<T>, Record<T>...]  -> Collect -> [Record<T>, Record<T>...] -> Collect...

Have now reset `delta_hash_map_` after `move`, so it would be always available for Record. Also added the test.
thanks for catching it.